--- xorg-x11.spec	2004-11-16 14:34:07.000000000 -0200
+++ xorg-x11.spec.elektra	2004-11-22 21:10:35.000000000 -0200
@@ -206,6 +206,9 @@
 
 %define rhbsys  %([ -r /etc/beehive-root ] && echo 1 || echo 0)
 
+# Weather to build the X server with support to Elektra instead of xorg.conf
+
+%define elektra                 1
 
 ######################################################################
 ######################################################################
@@ -492,7 +495,7 @@
 Patch20007: XFree86-4.3.0-redhat-xcursorgen-do-not-build-included-cursors.patch
 
 # EXPERIMENTAL patches (disabled except for development)
-
+Patch30000: xorg-x11-6.8.1-elektra.patch
 ######################################################################
 %description
 X.org X11 is an open source implementation of the X Window System.  It
@@ -1122,6 +1125,9 @@
 %patch20007 -p0 -b .redhat-xcursorgen-do-not-build-included-cursors
 %endif
 
+%if %{elektra}
+%patch30000 -p0 -b .elektra
+%endif
 
 ######################################################################
 #####  CREATE host-$arch.def  ##############################################
@@ -1148,6 +1154,7 @@
 
 # Copy our static architecture independant host.def into place
 cp %{SOURCE1} xc/config/cf/
+echo "#define UseElektra" >> xc/config/cf/host.def
 
 echo -e "\n\nGenerating architecture specifc host-%{_arch}.def file.\n\n"
 
