*** fsck.old.c	2004-12-27 12:41:07.321919712 +0100
--- fsck.c	2004-12-27 12:41:38.795135056 +0100
***************
*** 35,40 ****
--- 35,41 ----
  #include <ctype.h>
  #include <string.h>
  #include <time.h>
+ #include <kdb.h>
  #if HAVE_STDLIB_H
  #include <stdlib.h>
  #endif
***************
*** 63,68 ****
--- 64,71 ----
  #define	_PATH_MNTTAB	"/etc/fstab"
  #endif
  
+ #define KDB_FSTAB   "system/filesystems"
+ 
  static const char *ignored_types[] = {
  	"ignore",
  	"iso9660",
***************
*** 317,322 ****
--- 320,328 ----
  }
  
  /*
+  * Modified by Rémi Flament : added support for elektra api. (27/12/2004)
+  * If it can't find elektra keys it fallbacks to /etc/fstab and says it on the screen
+  *
   * Load the filesystem database from /etc/fstab
   */
  static void load_fs_info(const char *filename)
***************
*** 326,365 ****
  	int	lineno = 0;
  	int	old_fstab = 1;
  	struct fs_info *fs;
- 
- 	if ((f = fopen(filename, "r")) == NULL) {
- 		fprintf(stderr, _("WARNING: couldn't open %s: %s\n"),
- 			filename, strerror(errno));
- 		return;
- 	}
- 	while (!feof(f)) {
- 		lineno++;
- 		if (!fgets(buf, sizeof(buf), f))
- 			break;
- 		buf[sizeof(buf)-1] = 0;
- 		if (parse_fstab_line(buf, &fs) < 0) {
- 			fprintf(stderr, _("WARNING: bad format "
- 				"on line %d of %s\n"), lineno, filename);
- 			continue;
- 		}
- 		if (!fs)
- 			continue;
- 		if (fs->passno < 0)
- 			fs->passno = 0;
- 		else
- 			old_fstab = 0;
- 	}
  	
! 	fclose(f);
  	
! 	if (old_fstab) {
! 		fputs(_("\007\007\007"
! 		"WARNING: Your /etc/fstab does not contain the fsck passno\n"
! 		"	field.  I will kludge around things for you, but you\n"
! 		"	should fix your /etc/fstab file as soon as you can.\n\n"), stderr);
  		
! 		for (fs = filesys_info; fs; fs = fs->next) {
! 			fs->passno = 1;
  		}
  	}
  }
--- 332,429 ----
  	int	lineno = 0;
  	int	old_fstab = 1;
  	struct fs_info *fs;
  	
! 	KeySet *entries=ksNew();
! 	Key *current;
! 	int rc;
  	
! 	kdbOpen();
! 	rc=kdbGetChildKeys(KDB_FSTAB, entries, KDB_O_SORT|KDB_O_DIR);
! 	if (rc==0)
! 		{
! 		ksRewind(entries);
! 		while ((current=ksNext(entries))) 
! 			{
! 			char keyName[300];
! 			char buf[3];
! 			char *device;
! 			char *mpoint;
! 			char *type;
! 			char *opts;
! 			char *passno;
! 			char *freq;
! 			char *dev;
! 
! 			keyGetFullName(current,keyName,sizeof(keyName));
! 			
! 			// size of buffers are choosen arbitrary, I guess that should be changed...
! 			device=malloc(300);
! 			mpoint=malloc(300);
! 			type=malloc(100);
! 			opts=malloc(200);
! 			passno=malloc(2);
! 			freq=malloc(2);
! 			
! 			kdbGetValueByParent(keyName,"device",device,300);
! 			kdbGetValueByParent(keyName,"mpoint",mpoint,300);
! 			kdbGetValueByParent(keyName,"type",type,100);
! 			kdbGetValueByParent(keyName,"options",opts,200);
! 			kdbGetValueByParent(keyName,"passno",passno,3);
! 			kdbGetValueByParent(keyName,"dumpfreq",freq,3);
! 
! 			dev = blkid_get_devname(cache, device, NULL);
! 			if (dev)
! 				device = dev;
  		
! 			if (strchr(type, ','))
! 				type = 0;
! 		
! 			fs = create_fs_device(device, mpoint, type ? type : "auto", opts,
! 					freq ? atoi(freq) : -1,
! 					passno ? atoi(passno) : -1);
! 			if (dev)
! 				free(dev);
! 			
! 			
! 			}
! 		}
! 	else
! 		{
! 		printf("\nElektra keys haven't been found ! Falling back to /etc/fstab...\n\n");
! 		if ((f = fopen(filename, "r")) == NULL) {
! 			fprintf(stderr, _("WARNING: couldn't open %s: %s\n"),
! 				filename, strerror(errno));
! 			return;
! 		}
! 		while (!feof(f)) {
! 			lineno++;
! 			if (!fgets(buf, sizeof(buf), f))
! 				break;
! 			buf[sizeof(buf)-1] = 0;
! 			if (parse_fstab_line(buf, &fs) < 0) {
! 				fprintf(stderr, _("WARNING: bad format "
! 					"on line %d of %s\n"), lineno, filename);
! 				continue;
! 			}
! 			if (!fs)
! 				continue;
! 			if (fs->passno < 0)
! 				fs->passno = 0;
! 			else
! 				old_fstab = 0;
! 		}
! 		
! 		fclose(f);
! 		
! 		if (old_fstab) {
! 			fputs(_("\007\007\007"
! 			"WARNING: Your /etc/fstab does not contain the fsck passno\n"
! 			"	field.  I will kludge around things for you, but you\n"
! 			"	should fix your /etc/fstab file as soon as you can.\n\n"), stderr);
! 			
! 			for (fs = filesys_info; fs; fs = fs->next) {
! 				fs->passno = 1;
! 			}
  		}
  	}
  }
